enhancements/machine-config/admin-defined-reboot-policy.md

```diff
@@ -0,0 +1,264 @@
+      - type: file
+        value: /etc/containers/policy.json
+        actions:
+            - type: reload
+              target: crio.service
+      - type: file
+        value: /etc/containers/registries.conf
+        actions:
+            - type: special
+```
+
+Which, when applied, will define actions for all future changes to these MachineConfig fields based on the diff of the update (rendered MachineConfigs).
+
+For this initial implementation we will support MachineConfig changes to:
+ - Files
+ - Units
+ - kernelArguments
+ - sshKeys
+And actions for:
+ - None
+ - Draining a node
+ - Reloading a service
+ - Daemon-reload and restarting of a service
+ - Reboot (default)
+ - Special (see workflow description)
+
+In the future, we can extend this in two ways:
+Supporting more changes/actions (e.g. user defined script to run after a certain change)
+Supporting diffing between ostree images (for image based updates) and bootc managed systems (for bifrost)
+
+
+### Workflow Description
+
+An admin of an OCP cluster can at any time create/update/delete RebootPolicy objects to reconfigure how they wish updates to be applied. There isn’t any immediate effect so long as the RebootPolicy CR itself is a valid object.
+
+The MachineConfigController’s drain controller will process any update to the object and validate the RebootPolicy for duplication, action validity (but not correctness), etc. If the object is valid, the drain controller will update the object’s status and observedGeneration to indicate success, and if not, update those and degrade the controller, example:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "xxx"
+    Message: “”
+    status: "False"
+    type: RebootPolicyValidationFailing
+  - lastTransitionTime: "xxx"
+    Message: “Validation of the rebootpolicy has succeed.”
+    status: "True"
+    type: RebootPolicyValidated
+  observedGeneration: x
+```
+
+The MachineConfigDaemon will process and apply the current RebootPolicy if it’s valid, indicated by generation == observedGeneration (MCC has processed the latest spec) and RebootPolicyValidated == true/RebootPolicyValidationFailing == false. If not, the daemon will also degrade (and queue a re-sync) waiting for the controller to validate the object.
+
+For now, we only support one cluster-wide rebootpolicy, and as such any rebootpolicy not named “cluster” will be ignored and the MCO will report errors via degrading the operator.
+
+When the user first enables RebootPolicy via featuregate (or at some version, by default), the “cluster” RebootPolicy object will be generated with the existing MachineConfigDaemon defaults today, so the user can also view the default policies that came with the cluster. The user can update these fields if they wish to use a different workflow via userPolicies, but are not allowed to modify clusterDefaultPolicies, which is reserved for MCO to update these as we wish in the future. The fields that exist today are as highlighted above.
```

Comment:
How will you stop users modifying these, but allow the controller to update these? Perhaps this should be enforced by the "have a spec, copy to status, observe from status" pattern

Reply:
I intended to have the controller validate and overwrite any changes to those fields, much like any other object the MCO manages (for example, if you try to overwrite any of the controller-managed MachineConfig objects, the controller will simply re-apply the old one back. There is a delay so the rendered config doesn't get generated.)

The status will just reflect that the controller has ack'ed whatever is in the spec in this current pattern, so we can save re-listing all the policies in the status. But if the copy-to-status pattern is better, I think we can save a bit of waiting in the daemon (the current approach has the daemon wait for the latest, vs using the last-validated in the status in this new method) so it might be better?

Reply:
Ok, as an MVP, I think we can try for MachineConfiguration spec and status.